File: lib/connection.js
var sys = require("sys");
var Index = 0;
var CH = {};
var Waiting = {};
var Player = require("player");

var write = function(con, msg) {
  if (Waiting[con])
  {
    Waiting[con].send(msg);
    delete Waiting[con];
  }
  else
  {
    if (!CH[con]) CH[con] = [];
    CH[con].push(msg);
  }
}

Player.init(write);

exports.create = function() {
  CH[++Index] = [
    "Welcome to Baby Seal Node MUD.",
    "Make sure to bring a towel...",
    "type 'help' to begin!"
  ];
  return Index;
}

exports.process_command = function(con, msg) {
  var match = msg.trim().match(/^(\S+) *(.*)/);
  var cmd = match[1].toLowerCase();
  var arg = match[2].trim();
  if (arg.length > 0)
    Player.find(con).process_command(cmd, arg);
  else
    Player.find(con).process_command(cmd);
}

exports.flush = function(con, res) {
  if (CH[con] && CH[con].length > 0)
  {
    res.send(CH[con].join("\n"));
    CH[con] = [];
  }
  else
  {
    if (Waiting[con]) Waiting[con].end();
    Waiting[con] = res;
    CH[con] = [];
  }
}

File: lib/connection.rb
class Connection

  def initialize
    @id = 0;
    @cons = Hash.new []
    @waiting = {}
  end

  def create
    @id += 1
    @cons["#{@id}"] = [
      "Welcome to Baby Seal Node MUD.",
      "Make sure to bring a towel...",
      "type 'help' to begin!"
    ];
    @id
  end

  def flush(con, &blk)
    if (@cons[con].length > 0)
      blk.call(@cons[con].join("\n"));
      @cons[con] = [];
    else
      @waiting[con].call("") if @waiting[con]
      @waiting[con] = blk;
      @cons[con] = [];
    end
  end

  def write(con, msg)
    if (@waiting[con])
      @waiting[con].call(msg);
      @waiting.delete(con);
    else
      @cons[con] << msg;
    end
  end

  def process_command(con, cmd)
    ## add code here to parse the command and pass to the mud engine
    ## currently just echo back to the user
    write con, "ECHO: #{cmd}"
  end
end

